Encrypt password answers on config save

AlanCoding opened the pull request.

AlanCoding committed:
encrypt password answers on config save

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -3150,6 +3150,18 @@ def validate(self, attrs):
                         password_dict[key] = REPLACE_STR
                 if not self.instance or password_dict != self.instance.survey_passwords:
                     attrs['survey_passwords'] = password_dict
+                encrypt_dict(attrs['extra_data'], password_dict.keys())
```
https://github.com/ansible/awx/pull/846/files#diff-a81324c523b41de7296fdd5ff9063d10R3143

👀

ryanpetrello commented on awx/main/utils/encryption.py:
```diff
@@ -125,3 +127,13 @@ def decrypt_field(instance, field_name, subfield=None):
+    Encrypts all of the dictionary values in `data` under the keys in `fields`
+    in-place operation on `data`
+    '''
+    encrypt_fields = set(data.keys()) & set(fields)
```
Good choice w/ the `set`

For the less mathematically minded (me 😄 ), could we use:

`encrypt_fields = set(data.keys()).intersection(fields)`

I can never remember set notation and have to go looking to make sure I've got it right :)

ryanpetrello reviewed (approved):
Minus a few nits, this looks right to me.

AlanCoding committed:
validate against unencrypted values at spawn point

AlanCoding committed:
fix bug saving extra_data and follow prompts rules

display_extra_vars was not taking a copy of the data before
acting on it - this causes a bug where the activity stream will
modify the existing object on the model. That leads to new data
not being accepted.

Also moved the processing of extra_data to prior to the accept
or ignore kwargs logic so that we pass the right (post-encryption)
form of the variables.

AlanCoding committed:
allow WFJT nodes without required variables
